File: 6 - Expense Tracker/ExpenseTrackerFrontend/src/services/expenseService.js
import axios from "axios";

const BASE_URL = "http://localhost:8080";

export const getExpenses = () => axios.get(`${BASE_URL}/expenses`);

export const getExpense = (id) =>
    axios.get(`${BASE_URL}/expenses/${id}`);

export const addExpense = (expense) =>
    axios.post(`${BASE_URL}/expenses`, expense);

export const updateExpense = (id, expense) =>
    axios.put(`${BASE_URL}/expenses/${id}`, expense);

export const deleteExpense = (id) =>
    axios.delete(`${BASE_URL}/expenses/${id}`);

export const searchExpense = (title) =>
    axios.get(`${BASE_URL}/expenses/search?title=${title}`);

export const getCategoryExpense = () =>
    axios.get(`${BASE_URL}/dashboard/category`);

export const getMonthlyExpense = () =>
    axios.get(`${BASE_URL}/dashboard/monthly`);

export const getDashboardSummary = () =>
    axios.get(`${BASE_URL}/dashboard/summary`);

File: 6 - Expense Tracker/ExpenseTrackerBackend/src/main/java/com/example/ExpenseTrackerBackend/service/ExpenseService.java
package com.example.ExpenseTrackerBackend.service;

import com.example.ExpenseTrackerBackend.entity.Expense;
import com.example.ExpenseTrackerBackend.dto.CategoryExpenseDTO;
import com.example.ExpenseTrackerBackend.dto.DashboardSummary;
import com.example.ExpenseTrackerBackend.dto.MonthlyExpenseDTO;

import java.time.LocalDate;
import java.util.List;

public interface ExpenseService {

    // Add Expense
    Expense addExpense(Expense expense);

    // Get All Expenses
    List<Expense> getAllExpenses();

    // Get Expense By ID
    Expense getExpenseById(Long id);

    // Update Expense
    Expense updateExpense(Long id, Expense expense);

    // Delete Expense
    void deleteExpense(Long id);

    // Search by Title
    List<Expense> searchByTitle(String title);

    // Filter by Category
    List<Expense> getExpensesByCategory(String category);

    // Filter by Date
    List<Expense> getExpensesByDate(LocalDate date);

    // Filter by Date Range
    List<Expense> getExpensesByDateRange(LocalDate startDate, LocalDate endDate);

    // Dashboard Summary
    DashboardSummary getDashboardSummary();

    // Category Wise Expense
    List<CategoryExpenseDTO> getCategoryWiseExpense();

    // Monthly Expense
    List<MonthlyExpenseDTO> getMonthlyExpense();
}